File: 2-resources/__DATA-Structures/Code-Challenges/cc64quisksort/quicksort.c
/*
Implement the quick sort sorting algorithm.
Assume the input is an array of integers.

If you need to refresh your memory on the quick sort algorithm,
feel free to reference your old JavaScript implementation, as well as the following links:

https://en.wikipedia.org/wiki/Quicksort

https://www.khanacademy.org/computing/computer-science/algorithms#quick-sort
*/

#include <stdio.h>

void swap(int* a, int* b)
{
  int t = *a;
  *a = *b;
  *b = t;
}

int partition (int arr[], int low, int high)
{
  int pivot = arr[high];
  int i = (low - 1);

  for (int j = low; j <= high- 1; ++j)
  {
    if (arr[j] <= pivot)
    {
      ++i;
      swap(&arr[i], &arr[j]);
    }
  }
  swap(&arr[i + 1], &arr[high]);
  return (i + 1);
}

void quick_sort(int arr[], int low, int high)
{
  if (low < high)
  {
    int pi = partition(arr, low, high);
    quick_sort(arr, low, pi - 1);
    quick_sort(arr, pi + 1, high);
  }
}

void print_array(int arr[], int size)
{
  for (int i = 0; i < size; ++i) {
    printf("%d\n", arr[i]);
  }
}

int main(int argc, char* argv[])
{
  int arr[] = {100, 55, 4, 98, 10, 18, 90, 95, 43, 11, 47, 67, 89, 42, 49, 79};
  int n = sizeof(arr) / sizeof(arr[0]);
  quick_sort(arr, 0, n - 1);
  print_array(arr, n);
  return 0;
}

File: 2-resources/__DATA-Structures/Code-Challenges/cc22quickSort/quickSort.js
// cc22 quickSort
/*
 * Implement the quick sort sorting algorithm. Assume the input is an array of integers.
 * https://en.wikipedia.org/wiki/Quicksort
 * https://www.khanacademy.org/computing/computer-science/algorithms#quick-sort
 */
const quickSort = ( nums ) => {
  // BASE CASE
  if ( nums.length < 2 ) {
    return nums;
  } /* v */
  const pivot = nums[ 0 ],
    lessThan = [],
    greaterThan = [];
  /* v */
  for ( let i = 1; i < nums.length; i++ ) {
    if ( nums[ i ] < pivot ) {
      lessThan.push( nums[ i ] );
    } else greaterThan.push( nums[ i ] );
  };
  // Binary tree recursion? BigO((n^2)^2)
  return quickSort( lessThan ).concat( pivot, quickSort( greaterThan ) );
};


// TEST SUITE

// const x = [1,2,3],
//       y = 4
//       z = [5,6,7];
// console.log(x.concat(y, z)); // ~~~> [ 1, 2, 3, 4, 5, 6, 7 ]

console.log( quickSort( [ 9, 8, 7, 555, 632, 345, 4, 3, 2, 1 ] ) );
console.log( quickSort( [ 4, 6, 5, 3, 2 ] ) );
